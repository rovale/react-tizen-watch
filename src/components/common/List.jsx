import React from 'react';
import PropTypes from 'prop-types';

import './List.css';

export class List extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.object).isRequired,
    activeOptionId: PropTypes.string.isRequired,
    onActivateOption: PropTypes.func.isRequired,
    onSelectOption: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);

    this.onRotaryDetent = this.onRotaryDetent.bind(this);
    this.onSelectItem = this.onSelectItem.bind(this);
  }

  componentDidMount() {
    window.addEventListener('rotarydetent', this.onRotaryDetent);
  }

  componentWillUnmount() {
    window.removeEventListener('rotarydetent', this.onRotaryDetent);
  }

  onRotaryDetent(e) {
    const { direction } = e.detail;

    const index = this.props.options.findIndex(item => item.id === this.props.activeOptionId);

    if (direction === 'CW') {
      if (index < (this.props.options.length - 1)) {
        this.props.onActivateOption(this.props.options[index + 1].id);
      }
    } else if (index > 0) {
      this.props.onActivateOption(this.props.options[index - 1].id);
    }
  }

  onSelectItem(itemId) {
    this.props.onSelectOption(itemId);
  }

  render() {
    const { options } = this.props;

    const items = options.map(item => (
      <Item
        key={item.id}
        id={item.id}
        active={item.id === this.props.activeOptionId}
        onSelect={item.onSelect || this.onSelectItem}
      >
        {item.content || <div>{item.title}</div>}
      </Item>
    ));

    return (
      <ul className="ui-listview ui-snap-listview">
        {items}
      </ul>
    );
  }
}

export class Item extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    active: PropTypes.bool.isRequired,
    onSelect: PropTypes.func.isRequired,
    children: PropTypes.element.isRequired,
  };

  static contextTypes = {
    scrollToMiddle: PropTypes.func,
  };

  constructor(props) {
    super(props);

    this.scrollToMiddle = this.scrollToMiddle.bind(this);
    this.onSelect = this.onSelect.bind(this);
    this.element = null;
  }

  componentDidMount() {
    if (this.props.active) {
      window.setTimeout(() => this.scrollToMiddle(1), 0);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.active && !this.props.active) {
      window.setTimeout(() => this.scrollToMiddle(400), 0);
    }
  }

  onSelect(e) {
    e.preventDefault();
    this.props.onSelect(this.props.id);
  }

  scrollToMiddle(duration) {
    const itemRect = this.element.getBoundingClientRect();
    this.context.scrollToMiddle(duration, itemRect.top, itemRect.height);
  }

  render() {
    const { id, active, children } = this.props;
    return (
      <li ref={(e) => { this.element = e; }} className={`ui-snap-listview-item${active ? ' ui-snap-listview-selected' : ''}`}>
        <a href={`#${id}`} onClick={this.onSelect} id={id}>{children}</a>
      </li>
    );
  }
}
