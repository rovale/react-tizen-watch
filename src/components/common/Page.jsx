import React from 'react';
import PropTypes from 'prop-types';
import scroll from 'smooth-move';

export class Page extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  static childContextTypes = {
    scrollToMiddle: PropTypes.func,
  };

  constructor(props) {
    super(props);
    this.element = null;
  }

  getChildContext() {
    return { scrollToMiddle: this.scrollToMiddle };
  }

  scrollToMiddle = (duration, itemTop, itemHeight) => {
    const scroller = this.element.getElementsByClassName('ui-scroller')[0];
    const scrollerHeight = scroller.getBoundingClientRect().height;
    const newScrollTop = (itemTop + scroller.scrollTop) - ((scrollerHeight - itemHeight) / 2);

    scroll(scroller, { x: 0, y: newScrollTop, duration });
  }

  render() {
    return (
      <div ref={(e) => { this.element = e; }} className="ui-page ui-page-active ui-scroll-on">
        <div className="ui-scroller ui-snap-container" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
          {this.props.children}
        </div>
      </div>
    );
  }
}

export const Header = ({ children }) => (
  <header className="ui-header">
    <h2 className="ui-title">{children}</h2>
  </header>
);

Header.propTypes = {
  children: PropTypes.string.isRequired,
};

export const Content = ({ children }) => {
  return (
    <div className="ui-content" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
      {children}
    </div>);
};

Content.propTypes = {
  children: PropTypes.node.isRequired,
};
