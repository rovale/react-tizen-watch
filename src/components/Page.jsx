import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import * as action from '../actions/creators';
import { Page as UiPage, Header, Content } from './common/Page';
import { List } from './common/List';
import Switch from './deviceItem/Switch';
import Display from './deviceItem/Display';
import Presence from './deviceItem/Presence';
import Temperature from './deviceItem/Temperature';
import Thermostat from './deviceItem/Thermostat';

let Page = ({
  pageTitle, devices, activeDeviceId, onActivateDevice, onSelectDevice, onToggleSwitch,
}) => {
  const getListItemContent = (device) => {
    switch (device.template) {
      case 'device':
        return <Display device={device} />;
      case 'presence':
        return <Presence device={device} />;
      case 'switch':
        return <Switch device={device} />;
      case 'temperature':
        return <Temperature device={device} />;
      case 'thermostat':
        return <Thermostat device={device} />;
      default:
        return <div>{`${device.title} - (${device.template})`}</div>;
    }
  };

  const getOnSelect = (device) => {
    switch (device.template) {
      case 'switch':
        return () => onToggleSwitch(device.id);
      default:
        return null;
    }
  };

  const deviceOptions = devices.map(d => ({
    ...d,
    content: getListItemContent(d),
    onSelect: getOnSelect(d),
  }));

  return (
    <UiPage>
      <Header>{pageTitle}</Header>
      <Content>
        <List
          options={deviceOptions}
          activeOptionId={activeDeviceId}
          onActivateOption={onActivateDevice}
          onSelectOption={onSelectDevice}
        />
      </Content>
    </UiPage>);
};

Page.propTypes = {
  pageTitle: PropTypes.string.isRequired,
  devices: PropTypes.arrayOf(PropTypes.object).isRequired,
  activeDeviceId: PropTypes.string.isRequired,
  onActivateDevice: PropTypes.func.isRequired,
  onSelectDevice: PropTypes.func.isRequired,
  onToggleSwitch: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  const selectedPage = state.pages.find(p => p.id === state.ui.selectedPageId);

  return {
    pageTitle: selectedPage.title,
    devices: selectedPage.devices.map(pd => state.devices.find(d => pd === d.id)),
    activeDeviceId: state.ui.activeDeviceId || selectedPage.devices[0],
  };
};

const mapDispatchToProps = dispatch => ({
  onActivateDevice: id => dispatch(action.activateDevice(id)),
  onSelectDevice: id => dispatch(action.selectDevice(id)),
  onToggleSwitch: id => dispatch(action.toggleSwitch(id)),
});

export default Page = connect(mapStateToProps, mapDispatchToProps)(Page);
