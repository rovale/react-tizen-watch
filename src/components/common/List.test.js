import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { List, Item } from './List';
import { Page } from './Page';

Enzyme.configure({ adapter: new Adapter() });

describe('the List component', () => {
  let options;
  let onGenericSelect;
  let onSpecificSelect;
  let onActivate;

  let events;

  beforeEach(() => {
    events = {};
    window.addEventListener = jest.fn().mockImplementation((eventType, cb) => {
      events[eventType] = cb;
    });

    window.removeEventListener = jest.fn().mockImplementation((eventType) => {
      events[eventType] = null;
    });

    onGenericSelect = jest.fn();
    onSpecificSelect = jest.fn();
    onActivate = jest.fn();

    options = [
      { id: 'option1', title: 'Option 1' },
      { id: 'option2', content: <div><b>Option 2</b></div> },
      { id: 'option3', title: 'Option 3', onSelect: onSpecificSelect },
    ];
  });

  test('it should contain the provided items', () => {
    const list = mount(<List options={options} activeOptionId="option1" onActivateOption={() => {}} onSelectOption={() => {}} />);

    const items = list.find(Item);
    expect(items).toHaveLength(3);
    expect(list.find('ul')).toMatchSnapshot();
  });

  test('it should invoke the generic callback when an option without a specific callback is selected', () => {
    const list = mount(<List options={options} activeOptionId="option1" onActivateOption={() => {}} onSelectOption={onGenericSelect} />);
    const link = list.find('a[id="option1"]');

    link.simulate('click', { preventDefault: () => {} });

    expect(onGenericSelect.mock.calls).toHaveLength(1);
    expect(onGenericSelect.mock.calls[0][0]).toBe('option1');
    expect(onSpecificSelect.mock.calls).toHaveLength(0);
  });

  test('it should invoke the specific callback when an option with a specific callback is selected', () => {
    const list = mount(<List options={options} activeOptionId="option1" onActivateOption={() => {}} onSelectOption={onGenericSelect} />);
    const link = list.find('a[id="option3"]');

    link.simulate('click', { preventDefault: () => {} });

    expect(onGenericSelect.mock.calls).toHaveLength(0);
    expect(onSpecificSelect.mock.calls).toHaveLength(1);
    expect(onSpecificSelect.mock.calls[0][0]).toBe('option3');
  });

  test('it should activate the next item in the list when the bezel is turned clockwise', () => {
    mount(<List options={options} activeOptionId="option1" onActivateOption={onActivate} onSelectOption={() => {}} />);

    events.rotarydetent({ detail: { direction: 'CW' } });

    expect(onActivate.mock.calls).toHaveLength(1);
    expect(onActivate.mock.calls[0][0]).toBe('option2');
  });

  test('it should not go beyond the last item in the list when the bezel is turned clockwise', () => {
    mount(<List options={options} activeOptionId="option3" onActivateOption={onActivate} onSelectOption={() => {}} />);

    events.rotarydetent({ detail: { direction: 'CW' } });

    expect(onActivate.mock.calls).toHaveLength(0);
  });

  test('it should activate the previous item in the list when the bezel is turned counter clockwise', () => {
    mount(<List options={options} activeOptionId="option3" onActivateOption={onActivate} onSelectOption={() => {}} />);

    events.rotarydetent({ detail: { direction: 'CCW' } });

    expect(onActivate.mock.calls).toHaveLength(1);
    expect(onActivate.mock.calls[0][0]).toBe('option2');
  });

  test('it should not go before the first item in the list when the bezel is turned counter clockwise', () => {
    mount(<List options={options} activeOptionId="option1" onActivateOption={onActivate} onSelectOption={() => {}} />);

    events.rotarydetent({ detail: { direction: 'CCW' } });

    expect(onActivate.mock.calls).toHaveLength(0);
  });

  test('it should scroll, putting the active item in the middle', () => {
    jest.useFakeTimers();
    const page = mount(<Page><List options={options} activeOptionId="option3" onActivateOption={() => {}} onSelectOption={() => {}} /></Page>);
    const scroller = page.find('.ui-scroller').first().getDOMNode();
    jest.runAllTimers();

    expect(scroller.scrollTop).toBe(0);
  });

  test('it should cleanup the event listeners', () => {
    const list = mount(<List options={options} activeOptionId="option3" onActivateOption={() => {}} onSelectOption={() => {}} />);

    list.unmount();

    expect(events.rotarydetent).toBeNull();
  });
});

