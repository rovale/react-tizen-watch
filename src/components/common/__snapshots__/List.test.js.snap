// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the List component it should contain the provided items 1`] = `
<ul
  className="ui-listview ui-snap-listview"
>
  <Item
    active={true}
    id="option1"
    key="option1"
    onSelect={[Function]}
  >
    <li
      className="ui-snap-listview-item ui-snap-listview-selected"
    >
      <a
        href="#option1"
        id="option1"
        onClick={[Function]}
      >
        <div>
          Option 1
        </div>
      </a>
    </li>
  </Item>
  <Item
    active={false}
    id="option2"
    key="option2"
    onSelect={[Function]}
  >
    <li
      className="ui-snap-listview-item"
    >
      <a
        href="#option2"
        id="option2"
        onClick={[Function]}
      >
        <div>
          <b>
            Option 2
          </b>
        </div>
      </a>
    </li>
  </Item>
  <Item
    active={false}
    id="option3"
    key="option3"
    onSelect={[Function]}
  >
    <li
      className="ui-snap-listview-item"
    >
      <a
        href="#option3"
        id="option3"
        onClick={[Function]}
      >
        <div>
          Option 3
        </div>
      </a>
    </li>
  </Item>
</ul>
`;
